src/database/migrations/0055-create-audit-log.sql

```diff
@@ -0,0 +1,439 @@
+--
+-- PostgreSQL Audit Trigger Example
+-- Copyright (c) 2013, PostgreSQL Global Development Group
+--
+-- Permission to use, copy, modify, and distribute this software and its
+-- documentation for any purpose, without fee, and without a written agreement
+-- is hereby granted, provided that the above copyright notice and this
+-- paragraph and the following two paragraphs appear in all copies.
+--
+-- IN NO EVENT SHALL THE UNIVERSITY OF CALIFORNIA BE LIABLE TO ANY PARTY FOR
+-- DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES, INCLUDING
+-- LOST PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS
+-- DOCUMENTATION, EVEN IF THE UNIVERSITY OF CALIFORNIA HAS BEEN ADVISED OF THE
+-- POSSIBILITY OF SUCH DAMAGE.
+--
+-- THE UNIVERSITY OF CALIFORNIA SPECIFICALLY DISCLAIMS ANY WARRANTIES,
+-- INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
+-- AND FITNESS FOR A PARTICULAR PURPOSE.  THE SOFTWARE PROVIDED HEREUNDER IS
+-- ON AN "AS IS" BASIS, AND THE UNIVERSITY OF CALIFORNIA HAS NO OBLIGATIONS TO
+-- PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
+-- An audit history is important on most tables. Provide an audit trigger that
+-- logs to a dedicated audit table for the major relations.
+--
+-- This file should be generic and not depend on application roles or
+-- structures as it's being listed here:
+--
+--	https://wiki.postgresql.org/wiki/Audit_trigger_91plus
+--
+-- This trigger was originally based on
+--   http://wiki.postgresql.org/wiki/Audit_trigger
+-- but has been completely rewritten.
+--
+-- Should really be converted into a relocatable EXTENSION, with control and
+-- upgrade files.
+--
+-- Audited data. Lots of information is available, it's just a matter of how
+-- much you really want to record. See:
+--
+--   http://www.postgresql.org/docs/9.1/static/functions-info.html
+--
+-- Remember, every column you add takes up more audit table space and slows
+-- audit inserts.
+--
+-- Every index you add has a big impact too, so avoid adding indexes to the
+-- audit table unless you REALLY need them. The json GIN/GIST indexes are
+-- particularly expensive.
+--
+-- It is sometimes worth copying the audit table, or a coarse subset of it that
+-- you're interested in, into a temporary table where you CREATE any useful
+-- indexes and do your analysis.
+--
+CREATE TABLE audit_logs (
+	event_id bigserial PRIMARY KEY,
+	schema_name text NOT NULL,
+	table_name text NOT NULL,
+	relid oid NOT NULL,
+	session_user_name text,
+	action_tstamp_tx timestamp with time zone NOT NULL,
+	action_tstamp_stm timestamp with time zone NOT NULL,
+	action_tstamp_clk timestamp with time zone NOT NULL,
+	transaction_id bigint,
+	application_name text,
+	client_addr inet,
+	client_port integer,
+	client_query text,
+	action text NOT NULL CHECK (action IN ('I', 'D', 'U', 'T', 'A', 'S')),
+	row_data jsonb,
+	changed_fields jsonb,
+	statement_only boolean NOT NULL
+);
+REVOKE ALL ON audit_logs
```

Comment:
I know this is all boilerplate but in the spirit of knowing what's up -- do you know what this clause is accomplishing (the `REVOKE ... FROM public`).  Does this narrow which users are able to interact with audit_logs beyond what we have already configured?

If we eventually want to make these logs queryable via API will this mean our db user doesn't have access?

Reply:
It is worth taking a closer look, good call.

Before looking,  my rudimentary understanding is that this means only the owner and superuser will be able to use/see the table directly. At the moment, I think we use the same database user at runtime as migration-time, so there should be no issue querying it at runtime. Looking...

Reply:
So since the same user running migrations is the runtime user, I am confident that user should be able to query at runtime. I use the same user locally too, and it can access the table.

As for the practical effect of this `REVOKE` statement, in our case, it should be a no-op, meaning any existing users will not actually have access to new tables by default. I tried setting up a user like I had done in test and prod, and regardless of the presence of `REVOKE...` the user could not access the new `audit_logs` table.

Still, the `REVOKE` statement gives me warm fuzzies in the obscure case that we have some other non-admin user that for whatever reason has access to the `pdc` database.